src/tensors/cpu/wasm_intgemm_interface.h

```diff
@@ -0,0 +1,159 @@
+ * @param[in]   B_cols         No. of columns of input matrix. It should be a multiple of 8.
+ * @param[in]   cols           An array of column indices to be selected from input matrix.
+ *                             All indices of the array should be valid. i.e.
+ *                             i.e. 0 <= cols[N] < B_cols   where N = 0, 1, 2 .... (`num_cols`-1)
+ * @param[in]   num_cols       Size of the `cols` array. It should be a multiple of 8.
+ * @param[out]  output         An array representing the selected columns of input matrix.
+ *                             Size of the array = `B_rows` * `num_cols`s
+ */
+void Int8SelectColumnsB(const int8_t* B_input,
+                        Index B_rows,
+                        Index B_cols,
+                        const Index* cols,
+                        const Index num_cols,
+                        int8_t* output);
+
+
+/**
+ * Prepare A for the Matrix Multiply routine.
+ *
+ * It performs quantization on floating values.
+ *
+ * @param[in]   A_input     An array representing the input 2-D matrix in row-major format.
+ *                          Size of the array = `A_rows` * `A_cols`
+ * @param[in]   scale       The scaling factor (for quantization)
+ * @param[in]   zero_point  The zero point (for quantization)
+ * @param[in]   A_rows      No. of rows of input matrix. No restriction on its size.
+ * @param[in]   A_cols      No. of columns of input matrix. It should be a multiple of 64.
+ * @param[out]  output      An array representing the prepared A matrix in row-major format having
+ *                          the same size as of the input matrix i.e. `A_rows` * `A_cols`
+ */
+void Int8PrepareA(const float* A_input,
+                float scale,
+                int8_t zero_point,
+                Index A_rows,
+                Index A_cols,
+                int8_t* output);
+
+
+/**
+ * Prepares bias for the Matrix Multiply routine.
+ *
+ * @param[in]   B_input      An array representing the input 2-D matrix in row-major format from
+ *                           which bias will be prepared. Size of the array = `B_rows` * `B_cols`
+ * @param[in]   scale        The scaling factor (for quantization)
+ * @param[in]   zero_point   The zero point (for quantization)
+ * @param[in]   B_rows       No. of rows of input matrix. It should be a multiple of 64.
+ * @param[in]   B_cols       No. of columns of input matrix. It should be a multiple of 8.
+ * @param[in]   bias_input   An array representing the input bias having size = 1 * `B_cols`
+ * @param[out]  output       An array representing the prepared bias having size = 1 * `B_cols`
+ */
+void Int8PrepareBias(const int8_t* B_input,
+                    float scale,
+                    int8_t zero_point,
+                    Index B_rows,
+                    Index B_cols,
+                    const float* bias_input,
+                    float* output);
+
+
+/**
+ * A multiply routine to perform multiplication of 2 matrices.
```

Comment:
To avoid confusing others who may want to use this, we should point out that A_input is actually `uint8_t`, not `int8_t`.

Reply:
@XapaJIaMnu Wouldn't specifying this expose the architecture specific information in the interface? IIUC (and please correct me if I am wrong here):

1. The implementations of this interface for architectures that have a native `uint8 * int8` instruction (e.g. Intel and very recent Arm architecture) will do the following:
    1. `PrepareA` function will prepare A such that all of its entries are non-negative (e.g. by adding 127)
    2. `PrepareBias` will prepare a valid bias for the Multiply function
    3. `Multiply` function will perform multiplication internally using the corresponding native instruction (uint8 * int8)

2. The implementation of this interface for architectures that have a native `int8 * int8` instruction (e.g. Arm architecture) will do the following:
    1. `PrepareA` function will prepare A **_but it will not do anything to make all of its entries non-negative_**.
    2. `PrepareBias` will prepare **_an empty bias_** (as bias comes into picture only for supporting `uint8 * int8` case).
    3. `Multiply` function will perform multiplication internally using the corresponding native instruction (int8 * int8)

i.e. Most of the functions of this interface will have an architecture specific implementations but the user should not care about those details.

I can add in the documentation (wherever missing) that this interface implements things internally in architecture dependent manner. So the correct use of Multiply function requires using the corresponding `Prepare* `functions to prepare `A`, `B` and `bias`.

cc @kpu

Reply:
So, we always want to use `int8*int8` if this is available, as it simplifies everything massively, however this is not available on x86 (for any widely available hardware at least).

I think that we can add a function such as `isSignedTimesSigned()` that the client can use to determine the correct codepath. On top of that, we can also just support an identity `PrepareBias()` that does nothing.

Reply:
> So, we always want to use `int8*int8` if this is available, as it simplifies everything massively, however this is not available on x86 (for any widely available hardware at least).
> 

This is fine.

What I wanted to say is that we don't need to specify any implementation specific details in the interface (like you were saying here https://github.com/browsermt/marian-dev/pull/49#discussion_r685318208). User should be oblivious to them.

> I think that we can add a function such as `isSignedTimesSigned()` that the client can use to determine the correct codepath. On top of that, we can also just support an identity `PrepareBias()` that does nothing.

I believe we don't need this. AFAIR, spidermonkey will detect this on its own and call the right implementation. Right @yurydelendik  @eqrion ?

Reply:
I don't understand the question. The spidermonkey will not control any logic. If will just bridge the interface between wasm code and native (intrinsic) code. Whatever is implemented on either side will be your choice.

Reply:
> It will just bridge the interface between wasm code and native (intrinsic) code. Whatever is implemented on either side will be your choice.

As per my understanding, spidermonkey will determine which target architecture the code is running on and, based on this information, will hook the corresponding implementation of the matrix multiply. Please correct me if I am wrong here.

Reply:
I see intgemm is already doing cpuid. Unless you trust spidermonkey more to do that for the library, intgemm can keep doing it. Spidermonkey at this moment only detect if general platform is supported, e.g. ARM64 vs Intel, and enables intrinsics. Would you like spidermonkey to fo cpuid for you?

Reply:
> I think that we can add a function such as `isSignedTimesSigned()` that the client can use to determine the correct codepath. On top of that, we can also just support an identity `PrepareBias()` that does nothing.

@XapaJIaMnu Since we are providing one implementation of matrix multiply for each architecture (Intel SSSE3, AVX2, AVX512, Armv8.0, Armv8.4/8.5), wouldn't a cpu detection logic suffice to determine which implementation should be used under the hood? By `isSignedTimesSigned()`, you meant a cpu detection logic. Right?

Reply:
Currently if you're doing intel, you're always doing UnsignedTimesSigned, whereas if you're doing ARM you're doing SignedTimesSigned. It would be wasteful for the one implementing the multiplication to always force the presence of a bias so that the multiplication result is correct (Especially given that several big machine translation providers are moving away from biases for some of the matrices). We should allow users to choose a more efficient codepath when possible.

Reply:
For this, I believe the solution is pretty straightforward. I can add one more Multiply function that doesn't take any bias argument.
In all, there will exist 2 Multiply functions:

1. `MultiplyAndAddBias(A, B, Bias, output)`
    1. It will do `unsigned * signed` multiplication by using the trick `A * B + Bias`
    2. It will require a valid implementation of `PrepareBias`
2. `Multiply(A, B, output)`
    1. It will do `signed * signed` multiplication by just `A * B`
    2. It doesn't require `PrepareBias`

However, the 2nd `Multiply` function can be added later as well when you provide the Arm implementation.

Reply:
I changed the name of the existing `Multiply` function to `MultiplyAndAddBias` and added documentation that this function can be used for supporting matrix multiplication on architectures that don't have a native instruction for signed*signed but have a native instruction for unsigned*signed

Reply:
@XapaJIaMnu I probably don't understand what the issue is here. If the concern is that how will we support both `signed * signed` and `unsigned * signed` then I see 2 options:

1. We support both cases using single Multiply function. This means:
    1. For architectures having only `unsigned * signed` native instruction (e.g. Intel)
        1. `PrepareA` will prepare A so that every value is non-negative and
        2. `PrepareBias` prepares a valid bias for the `Multiply` function
    2. For architectures having `signed * signed` native instruction (e.g. ARM)
        1. `PrepareA` will not try to make every value non-negative
        2. `PrepareBias` prepares an empty bias (or identity bias as you were saying)
2. We support both cases using 2 different Multiply functions:
    1. For `signed * signed`, something like `Multiply(A, B)`
        1. This requires implementing only PrepareA and PrepareB 
    2. For `unsigned * signed`, something like `MultiplyAndAddBias(A, B, Bias)`
        1. This requires implementing PrepareA, PrepareB and PrepareBias

During last plenary meeting we decided that Option 1 is a satisfactory approach. Yes, in this option, I agree with you that the architectures that support `signed * signed` will be forced to provide implementation of `PrepareBias` even though bias is not needed for multiply.

Do you have some other concern that I am missing here?

Reply:
I'm debating a system like this:

B and bias are always carried together in an opaque blob.  There's a `void PackBAndBias(int8_t *blob, const float *B, const float *bias, Index B_rows, Index B_cols, float scaling)` that allows the bias term to be `nullptr` for no bias.  The blob would be variable size (it might create a bias term or not; it might be padded), so there's another call to tell the user how much memory it needs `size_t SizeBAndBias(Index B_rows, Index B_cols, bool have_bias)`.  

A is packed in an opaque blob.  There's a `void PackA(int8_t *blob, const float *A, Index A_rows, Index B_cols, float scaling)` and maybe an analogous `size_t SizeA(Index A_rows, Index A_cols)` though we don't need one now because it's just A_rows * A_cols.  

One Multiply function gets PackBAndBias's blob and PackA's blob.  It already knows if there is bias because PackB put it in the opaque blob.  On Intel there always will be bias.  If Nick wants to do ARM without bias, he can.  

There's also a flavor of Multiply that goes directly from float for A.

Reply:
@kpu @XapaJIaMnu Please let me know your thoughts on the last comment 👍

Reply:
I think we should go ahead without the fanciness of not having bias and variable sizing.  Then only if it's justified by the speed it can be extended.

Reply:
Okay. I will leave everything as it is then 👍